ccl/sqlproxyccl: TestDirectoryConnect failed

ccl/sqlproxyccl.TestDirectoryConnect [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16502823?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16502823?buildTab=artifacts#/) on release-24.1 @ [5066975f872cd6a452fbe8306befa75c4ff5132e](https://github.com/cockroachdb/cockroach/commits/5066975f872cd6a452fbe8306befa75c4ff5132e):


```
=== RUN   TestDirectoryConnect
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/c5841474bf5b4c6fc7dfbf4575fb8282/logTestDirectoryConnect1912160775
    test_log_scope.go:81: use -show-logs to present logs inline
    proxy_handler_test.go:1450: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c5841474bf5b4c6fc7dfbf4575fb8282/logTestDirectoryConnect1912160775
--- FAIL: TestDirectoryConnect (48.63s)
=== RUN   TestDirectoryConnect/successful_connection_after_restart
    proxy_handler_test.go:1441: condition failed to evaluate within 45s: failed to connect to ``host=127.0.0.1 user=testuser database=``: server error (ERROR: server not configured to accept remote address override (requested: "127.0.0.1:58242") (SQLSTATE 08P01))
    --- FAIL: TestDirectoryConnect/successful_connection_after_restart (45.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDirectoryConnect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-41397

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -1894,6 +1894,10 @@ func TestPodWatcher(t *testing.T) {
 	opts.testingKnobs.balancerOpts = []balancer.Option{
 		balancer.NoRebalanceLoop(),
 		balancer.RebalanceRate(1.0),
+		// Set a rebalance delay of zero. Rebalance is triggered on startup because
+		// the watch enumerates three existing pods. The delay of zero allows
+		// addition of the fourth pod to trigger rebalancing.
+		balancer.RebalanceDelay(0),
 	}
 	proxy, addrs := newSecureProxyServer(ctx, t, s.Stopper(), opts)
 	connectionString := fmt.Sprintf("postgres://testuser:hunter2@%s/?sslmode=require&options=--cluster=tenant-cluster-%s", addrs.listenAddr, tenantID)

--- a/pkg/ccl/sqlproxyccl/tenantdirsvr/BUILD.bazel
+++ b/pkg/ccl/sqlproxyccl/tenantdirsvr/BUILD.bazel
@@ -16,6 +16,7 @@ go_library(
         "//pkg/roachpb",
         "//pkg/util/grpcutil",
         "//pkg/util/log",
+        "//pkg/util/protoutil",
         "//pkg/util/stop",
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",

--- a/pkg/ccl/sqlproxyccl/tenantdirsvr/test_static_directory_svr.go
+++ b/pkg/ccl/sqlproxyccl/tenantdirsvr/test_static_directory_svr.go
@@ -15,6 +15,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -162,6 +163,19 @@ func (d *TestStaticDirectoryServer) WatchPods(
 	c := make(chan *tenant.WatchPodsResponse, 10)
 	chElement := addListener(c)
 
+	initialPods := func() (result []tenant.WatchPodsResponse) {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		for _, pods := range d.mu.tenantPods {
+			for _, pod := range pods {
+				result = append(result, tenant.WatchPodsResponse{
+					Pod: protoutil.Clone(pod).(*tenant.Pod),
+				})
+			}
+		}
+		return result
+	}()
+
 	return stopper.RunTask(
 		server.Context(),
 		"watch-pods-server",
@@ -172,19 +186,24 @@ func (d *TestStaticDirectoryServer) WatchPods(
 				}
 			}()
 
-			for watch := true; watch; {
+			for i := range initialPods {
+				if err := server.Send(&initialPods[i]); err != nil {
+					return
+				}
+			}
+
+			for {
 				select {
 				case e, ok := <-c:
 					// Channel was closed.
 					if !ok {
-						watch = false
-						break
+						return
 					}
 					if err := server.Send(e); err != nil {
-						watch = false
+						return
 					}
 				case <-stopper.ShouldQuiesce():
-					watch = false
+					return
 				}
 			}
 		},
@@ -274,6 +293,18 @@ func (d *TestStaticDirectoryServer) WatchTenants(
 	c := make(chan *tenant.WatchTenantsResponse, 10)
 	chElement := addListener(c)
 
+	initialTenants := func() (result []tenant.WatchTenantsResponse) {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		for id := range d.mu.tenants {
+			result = append(result, tenant.WatchTenantsResponse{
+				Type:   tenant.EVENT_ADDED,
+				Tenant: protoutil.Clone(d.mu.tenants[id]).(*tenant.Tenant),
+			})
+		}
+		return result
+	}()
+
 	return stopper.RunTask(
 		server.Context(),
 		"watch-tenants-server",
@@ -284,19 +315,24 @@ func (d *TestStaticDirectoryServer) WatchTenants(
 				}
 			}()
 
-			for watch := true; watch; {
+			for i := range initialTenants {
+				if err := server.Send(&initialTenants[i]); err != nil {
+					return
+				}
+			}
+
+			for {
 				select {
 				case e, ok := <-c:
 					// Channel was closed.
 					if !ok {
-						watch = false
-						break
+						return
 					}
 					if err := server.Send(e); err != nil {
-						watch = false
+						return
 					}
 				case <-stopper.ShouldQuiesce():
-					watch = false
+					return
 				}
 			}
 		},
@@ -561,12 +597,10 @@ func (d *TestStaticDirectoryServer) notifyTenantUpdateLocked(
 	typ tenant.WatchEventType, t *tenant.Tenant,
 ) {
 	// Make a copy of the tenant to prevent race issues.
-	copyTenant := *t
-	copyTenant.AllowedCIDRRanges = make([]string, len(t.AllowedCIDRRanges))
-	copy(copyTenant.AllowedCIDRRanges, t.AllowedCIDRRanges)
-	copyTenant.AllowedPrivateEndpoints = make([]string, len(t.AllowedPrivateEndpoints))
-	copy(copyTenant.AllowedPrivateEndpoints, t.AllowedPrivateEndpoints)
-	res := &tenant.WatchTenantsResponse{Type: typ, Tenant: &copyTenant}
+	res := &tenant.WatchTenantsResponse{
+		Type:   typ,
+		Tenant: protoutil.Clone(t).(*tenant.Tenant),
+	}
 
 	for e := d.mu.tenantEventListeners.Front(); e != nil; {
 		select {

